wallet/wallet.c

```diff
@@ -3017,15 +3039,79 @@ struct channeltx *wallet_channeltxs_get(struct wallet *w, const tal_t *ctx,
 	return res;
 }
 
+static bool wallet_forwarded_payment_update(struct wallet *w,
+					    const struct htlc_in *in,
+					    const struct htlc_out *out,
+					    enum forward_status state,
+					    enum onion_type failcode,
+					    struct timeabs *resolved_time)
+{
+	struct db_stmt *stmt;
+	bool changed;
+
+	/* We update based solely on the htlc_in since an HTLC cannot be
+	 * associated with more than one forwarded payment. This saves us from
+	 * having to have two versions of the update statement (one with and
+	 * one without the htlc_out restriction).*/
+	stmt = db_prepare_v2(w->db,
+			     SQL("UPDATE forwarded_payments SET"
+				 "  in_msatoshi=?"
+				 ", out_msatoshi=?"
+				 ", state=?"
+				 ", resolved_time=?"
+				 ", failcode=?"
+				 " WHERE in_htlc_id=?"));
+	db_bind_amount_msat(stmt, 0, &in->msat);
+
+	if (out) {
+		db_bind_amount_msat(stmt, 1, &out->msat);
+	} else {
+		db_bind_null(stmt, 1);
+	}
+
+	db_bind_int(stmt, 2, wallet_forward_status_in_db(state));
+
+	if (resolved_time != NULL) {
+		db_bind_timeabs(stmt, 3, *resolved_time);
+	} else {
+		db_bind_null(stmt, 3);
+	}
+
+	if(failcode != 0) {
+		assert(state == FORWARD_FAILED || state == FORWARD_LOCAL_FAILED);
+		db_bind_int(stmt, 4, (int)failcode);
+	} else {
+		db_bind_null(stmt, 4);
+	}
+
+	db_bind_u64(stmt, 5, in->dbid);
+	db_exec_prepared_v2(stmt);
+	changed = db_count_changes(stmt) != 0;
+	tal_free(stmt);
+
+	return changed;
+}
+
 void wallet_forwarded_payment_add(struct wallet *w, const struct htlc_in *in,
 				  const struct htlc_out *out,
 				  enum forward_status state,
 				  enum onion_type failcode)
 {
 	struct db_stmt *stmt;
 	struct timeabs *resolved_time;
+
+	if (state == FORWARD_SETTLED || state == FORWARD_FAILED) {
+		resolved_time = tal(tmpctx, struct timeabs);
+		*resolved_time = time_now();
+	} else {
+		resolved_time = NULL;
+	}
+
+	if (wallet_forwarded_payment_update(w, in, out, state, failcode, resolved_time))
+		goto notify;
+
 	stmt = db_prepare_v2(w->db,
-			     SQL("INSERT OR REPLACE INTO forwarded_payments ("
+			     SQL("INSERT INTO forwarded_payments ("
 				 "  in_htlc_id"
```

Comment:
Upsert went into postgres 9.5, and we seem to be insisting on postgres 10 anyway?

Reply:
Yep, but the `ON CONFLICT(field) DO UPDATE` syntax is not in `sqlite3` below 3.24, so we can't use it. Splitting it into two statements was the only really portable way I found.

Reply:
Yech.  Nobody cares about compatibility, do they?!?